src/datachain/query/dataset.py

```diff
@@ -957,24 +957,24 @@ def q(*columns):
 
 
 @frozen
-class GroupBy(Step):
-    """Group rows by a specific column."""
+class SQLGroupBy(SQLClause):
+    cols: Sequence[Union[str, ColumnElement]]
+    group_by: Sequence[Union[str, ColumnElement]]
 
-    cols: PartitionByType
-
-    def clone(self) -> "Self":
-        return self.__class__(self.cols)
+    def apply_sql_clause(self, query) -> Select:
+        if not self.cols:
+            raise ValueError("No columns to select")
+        if not self.group_by:
+            raise ValueError("No columns to group by")
 
-    def apply(
-        self, query_generator: QueryGenerator, temp_tables: list[str]
-    ) -> StepResult:
-        query = query_generator.select()
-        grouped_query = query.group_by(*self.cols)
+        subquery = query.subquery()
 
-        def q(*columns):
-            return grouped_query.with_only_columns(*columns)
+        cols = [
+            subquery.c[str(c)] if isinstance(c, (str, C)) else c
```

Comment:
Any reason why `query.selected_columns` won't work here instead of a subquery?

Reply:
The reason is if we are using `LIMIT` in the query, `group_by` will only work within the limited amount of columns. That's why I've added subquery here, to be sure limit will not affect aggregation.

Reply:
Are you sure that limit is applied before group by in SQL? I thought it was applied after group by. I.e 

```SQL

-- create
CREATE TABLE EMPLOYEE_SALES (
  empId INTEGER NOT NULL,
  name TEXT NOT NULL,
  dept TEXT NOT NULL,
  sale_value FLOAT NOT NULL
);

-- insert
INSERT INTO EMPLOYEE_SALES VALUES (0001, 'Clark', 'Sales', 10);
INSERT INTO EMPLOYEE_SALES VALUES (0001, 'Clark', 'Sales', 10);
INSERT INTO EMPLOYEE_SALES VALUES (0001, 'Clark', 'Sales', 300);
INSERT INTO EMPLOYEE_SALES VALUES (0001, 'Clark', 'Sales',400);
INSERT INTO EMPLOYEE_SALES VALUES (0001, 'Clark', 'Sales',1);
INSERT INTO EMPLOYEE_SALES VALUES (0001, 'Clark', 'Sales',5);
INSERT INTO EMPLOYEE_SALES VALUES (0002, 'Dave', 'Accounting', 20);
INSERT INTO EMPLOYEE_SALES VALUES (0003, 'Ava', 'Sales',400);

-- fetch 
SELECT 
  count(*) as count, 
  sum(sale_value) as total_sales, 
  name 
FROM EMPLOYEE_SALES group by name limit 3;
```
yields:

```console
Output:

1|400.0|Ava
6|726.0|Clark
1|20.0|Dave
```

Reply:
Let me check again, but I've added subquery because of the limit issue (after using `.show()` chain method). I'll be back with the answer.

Reply:
I can not find the way to reproduce issue with limit in current implementation, but removing subquery causes another issues with SQL functions and labels.
My suggestion is to leave it as is for now, because it is another issue — to get rid of subqueries. Right now they are actively used in all steps, even [basic select](https://github.com/iterative/datachain/blob/main/src/datachain/query/dataset.py#L709-L719).

I have some work done on this, removed almost all subqueries and in the end there is non performance boost without subqueries in both CLI and SaaS, but there was some issues with tests. Also @rlamy was looking into this. I can create a follow-up issue to solve subqueries issue.